os/kernel/pthread/pthread_mutexunlock.c

```diff
@@ -117,16 +146,24 @@
 
 int pthread_mutex_unlock(FAR pthread_mutex_t *mutex)
 {
-	int ret = EINVAL;
+	int ret = OK;
 
 	svdbg("mutex=0x%p\n", mutex);
 	DEBUGASSERT(mutex != NULL);
+	if (mutex == NULL) {
+		return EINVAL;
+	}
 
 	/* Make sure the semaphore is stable while we make the following checks.
 	 * This all needs to be one atomic action.
 	 */
 	sched_lock();
-	if (mutex != NULL) {
+
+	/* The unlock operation is only performed if the mutex is actually locked.
+	 * If the mutex is not locked, then SUCCESS will be returned (there is
+	 * no error return value specified for this case).
+	 */
+	if (pthread_mutex_islocked(mutex)) {
```

Comment:
Ok Fine.  We can return ERROR.   And it's surprising to see that most ( 90 % ) of the application codes does not handle the return statement of pthread_mutex_lock or pthread_mutexunlock .  It's very dangerous situation as  pthread_mutexlock / unlock would fail for so many reasons.  Without checking the SUCCESS case, they enter into critical section.

I think we should take a serious note on this